extension/goplugin/types_test.go

```diff
@@ -0,0 +1,185 @@
+			Entry("null vs undefined", goext.MakeNullString(), goext.MakeUndefinedString(), true),
+			Entry("null vs defined", goext.MakeNullString(), goext.MakeString("hello"), false),
+			Entry("defined vs defined", goext.MakeString("hello"), goext.MakeString("hello"), true),
+			Entry("defined vs defined different value", goext.MakeString("hello"), goext.MakeString("world"), false),
+		)
+
+		DescribeTable("Float",
+			func(lhv goext.MaybeFloat, rhv goext.MaybeFloat, expected bool) {
+				Expect(lhv.Equals(rhv)).To(Equal(expected))
+			},
+			Entry("undefined vs undefined", goext.MakeUndefinedFloat(), goext.MakeUndefinedFloat(), true),
+			Entry("undefined vs defined", goext.MakeUndefinedFloat(), goext.MakeFloat(1.23), false),
+			Entry("null vs undefined", goext.MakeNullFloat(), goext.MakeUndefinedFloat(), true),
+			Entry("null vs defined", goext.MakeNullFloat(), goext.MakeFloat(1.23), false),
+			Entry("defined vs defined", goext.MakeFloat(1.23), goext.MakeFloat(1.23), true),
+			Entry("defined vs defined different value", goext.MakeFloat(1.23), goext.MakeFloat(3.21), false),
+		)
+
+		DescribeTable("Int",
+			func(lhv goext.MaybeInt, rhv goext.MaybeInt, expected bool) {
+				Expect(lhv.Equals(rhv)).To(Equal(expected))
+			},
+			Entry("undefined vs undefined", goext.MakeUndefinedInt(), goext.MakeUndefinedInt(), true),
+			Entry("undefined vs defined", goext.MakeUndefinedInt(), goext.MakeInt(123), false),
+			Entry("null vs undefined", goext.MakeNullInt(), goext.MakeUndefinedInt(), true),
+			Entry("null vs defined", goext.MakeNullInt(), goext.MakeInt(123), false),
+			Entry("defined vs defined", goext.MakeInt(123), goext.MakeInt(123), true),
+			Entry("defined vs defined different value", goext.MakeInt(123), goext.MakeInt(321), false),
+		)
+
+		DescribeTable("Bool",
+			func(lhv goext.MaybeBool, rhv goext.MaybeBool, expected bool) {
+				Expect(lhv.Equals(rhv)).To(Equal(expected))
+			},
+			Entry("undefined vs undefined", goext.MakeUndefinedBool(), goext.MakeUndefinedBool(), true),
+			Entry("undefined vs defined", goext.MakeUndefinedBool(), goext.MakeBool(true), false),
+			Entry("null vs undefined", goext.MakeNullBool(), goext.MakeUndefinedBool(), true),
+			Entry("null vs defined", goext.MakeNullBool(), goext.MakeBool(true), false),
+			Entry("defined vs defined", goext.MakeBool(true), goext.MakeBool(true), true),
+			Entry("defined vs defined different value", goext.MakeBool(true), goext.MakeBool(false), false),
+		)
+	})
+
+	Describe("JSON Marshalling", func() {
+		Context("String", func() {
+			type TestResource struct {
+				Value goext.MaybeString `json:"value"`
+			}
+			It("value defined", func() {
+				buf, err := json.Marshal(TestResource{Value: goext.MakeString("hello")})
+				Expect(err).To(BeNil())
+				Expect(string(buf)).To(Equal(`{"value":"hello"}`))
+			})
+
+			It("value undefined", func() {
+				buf, err := json.Marshal(TestResource{})
+				Expect(err).To(BeNil())
+				Expect(string(buf)).To(Equal(`{"value":null}`))
```

Comment:
Please correct me if I'm wrong. As far as I remember, undefined value is omitted with key in JS extension. That is, if there is `{"key1": null, "key2": undefined, "key3": "string"}`, the result JSON will be `{"key1": null, "key3": "string"}`.

Reply:
Yes, it should be interpreted like that. Unfortunately json.Marshaller and json.Unmarshaller interface AFAIK doesn't support this case. That's why we should always use env.Util().Resource(To/From)Map and then marshal return map to JSON. These tests just cover how UnmarshalJSON and MarshalJSON functions are implemented on this types. This case but when using env.Util().Resource(To/From)Map are covered in utils_test.go.

Reply:
Got it. That makes sense.